lib/tasks/one_off.rake

```diff
@@ -118,9 +118,59 @@ namespace :one_off do
     puts "Updating #{projects.count} projects..."
 
     projects.in_batches(of: 10000).each do |batch|
-      batch.update_all('status_checked_at = updated_at')
+      batch.update_all("status_checked_at = updated_at")
     end
 
     print "Finished updating projects"
   end
+
+  desc "Backfill Pypi project dependencies kind with environment markers"
+  task :backfill_pypi_dependencies_kind, %i[batch_size] => :environment do |_t, args|
+    # There's ~5 million Pypi versions
+    # Based on a sampling of 5000 Pypi versions, ~1/10 contain environment markers
+    # 4 Pypi API calls per second = 240/min
+    # leaving a buffer of 40 for any batches which contain more, 200/min is conservative.
+    # 200/min = 12000/hour = 288k/day
+    # At that rate it will take ~1.75 days
+
+    pypi_versions = Version.where(
+      project_id: Project.where(platform: "Pypi")
+    )
+
+    # works out to ~200 affected versions, 4 per second, with room for overflow so api-calls don't get too stacked
+    # on each other if they overflow into the next minute.
+    batch_size = args[:batch_size] || 2000
+
+    num_batches = pypi_versions.count / batch_size
+
+    environment_markers = PackageManager::Pypi::PEP_508_ENVIRONMENT_MARKERS
+      .split("|")
+      .map { |em| "%#{em}%" }
+      .join(", ")
+
+    pypi_versions.in_batches(of: batch_size).each_with_index do |batch_versions, batch_versions_index|
+      affected_versions = Dependency
+        .where(version: batch_versions)
+        .joins(version: :project)
+        .where("dependencies.requirements LIKE any (array[#{environment_markers}])")
+        .select("versions.number as project_version, projects.name as project_name")
```

Comment:
How's the performance on this query? Looks like there is an index on `dependencies.version_id`, so that's good. We could potentially offload everything in this block to a new worker

Reply:
I actually had this query completely wrong. It was duplicating versions, once for each dependency with environment markers. I think the performance of this query on a batch is ok. I made it so we can pass in batch size and the start so we can tweak the running of it as needed. I think putting in a `sleep` would have essentially the same effect as moving it to its own worker no? Moving it to its own worker seems like a lot of work for a one-off so I'd like to avoid it if I can.